### PR TITLE
Task/mongo repo

### PR DESCRIPTION
**Pull request by alecc08**
Sorry for mega PR 😆 lots of features here...

- Mongo keeping track of coconut giving/receiving/daily limits
- Lots of refactoring of speech code (sending messages to channels)
- Bug fixes..

### CODE REVIEW

File: src/main/java/org/noixdecoco/app/rest/MainREST.java [hunk 1]
```diff
@@ -67,33 +57,41 @@ public void insertPerson(@RequestParam("id") Long id) {
...
-						String data = "{ \"channel\":\""+event.getEvent().getChannel()+"\", \"text\": \"DID <@"+event.getEvent().getUser()  + "> just give coconuts to ";
+						String text = "";
 						for(String name : names) {
-							data += "<@" + name + " ";
+							try {
+								long numCoconuts = coconutService.addCoconut(event.getEvent().getUser(), name, numberOfCoconuts);
+								text += "<@"+event.getEvent().getUser()  + "> gave " + numberOfCoconuts +
+										" coconut" + (numberOfCoconuts > 1 ? "s":"") + " to <@" + name + ">, " +
+										" they now have " + numCoconuts + " coconut" + (numCoconuts > 1 ? "s":"") + ". ";
+							}  catch (InsufficientCoconutsException e) {
```
Comment (RaymondAudet) [inline src/main/java/org/noixdecoco/app/rest/MainREST.java]: reformat code!!

File: src/main/java/org/noixdecoco/app/service/impl/CoconutServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	@Autowired
+	private CoconutLedgerRepository coconutRepo;
+	
+	@Value("${daily.coconut.limit}")
+	private long dailyLimit;
+
+	@Override
+	public long addCoconut(String fromUser, String toUser, int numCoconuts) throws CoconutException {
+		if(fromUser.equals(toUser)) {
+			throw new InvalidReceiverException();
```
Comment (RaymondAudet) [inline src/main/java/org/noixdecoco/app/service/impl/CoconutServiceImpl.java]: exceptions as flow-control? Could we instead validate the request before and leave exceptions for real problems?

File: src/main/java/org/noixdecoco/app/service/impl/CoconutServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+@Service
+public class CoconutServiceImpl implements CoconutService {
+	
+	private static final Logger LOGGER = LogManager.getLogger(CoconutService.class);
+	
+	@Autowired
+	private CoconutLedgerRepository coconutRepo;
+	
+	@Value("${daily.coconut.limit}")
+	private long dailyLimit;
```
Comment (RaymondAudet) [inline src/main/java/org/noixdecoco/app/service/impl/CoconutServiceImpl.java]: at which point do we set back the daily limit of coconuts to users? at 12:01 each night to everybody or only if one tries to give one and he hasn't given any today?

Comment (alecc08) [reply src/main/java/org/noixdecoco/app/service/impl/CoconutServiceImpl.java]: When giving a coconut, the code checks when the last time a coconut was given, if it was given before the start of today, the coconuts given resets to 0

File: src/main/java/org/noixdecoco/app/rest/MainREST.java (hunk 1 above)
Comment (alecc08) [reply src/main/java/org/noixdecoco/app/rest/MainREST.java]: Good point but I'd want to keep the validation code inside the service. Could switch to returning an enumeration of possible statuses though.

**Pull request merged**